apps/OboeTester/app/src/main/cpp/ReverseJniEngine.cpp

```diff
@@ -0,0 +1,126 @@
+    env->DeleteGlobalRef(mJavaObject);
+}
+
+void ReverseJniEngine::start(int bufferSizeInBursts) {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "Starting audio stream.");
+    setupAudioStream();
+    setBufferSizeInBursts(bufferSizeInBursts);
+    if (mAudioStream) {
+        mAudioStream->requestStart();
+    }
+}
+
+void ReverseJniEngine::stop() {
+    __android_log_print(ANDROID_LOG_INFO, TAG, "Stopping audio stream.");
+    if (mAudioStream) {
+        mAudioStream->requestStop();
+        closeAudioStream();
+    }
+}
+
+void ReverseJniEngine::setBufferSizeInBursts(int bufferSizeInBursts) {
+    if (mAudioStream) {
+        mAudioStream->setBufferSizeInFrames(bufferSizeInBursts * mAudioStream->getFramesPerBurst());
+        __android_log_print(ANDROID_LOG_INFO, TAG, "Buffer size set to %d frames.", mAudioStream->getBufferSizeInFrames());
+    }
+}
+
+void ReverseJniEngine::setupAudioStream() {
+    oboe::AudioStreamBuilder builder;
+    builder.setDirection(oboe::Direction::Output)
+            ->setPerformanceMode(oboe::PerformanceMode::LowLatency)
+            ->setSharingMode(oboe::SharingMode::Exclusive)
+            ->setFormat(oboe::AudioFormat::Float)
+            ->setChannelCount(oboe::ChannelCount::Mono)
+            ->setDataCallback(this);
+
+    oboe::Result result = builder.openStream(mAudioStream);
+    if (result != oboe::Result::OK) {
+        __android_log_print(ANDROID_LOG_ERROR, TAG, "Failed to create stream. Error: %s", oboe::convertToText(result));
+    }
+}
+
+void ReverseJniEngine::closeAudioStream() {
+    if (mAudioStream) {
+        mAudioStream->close();
+    }
+}
+
+oboe::DataCallbackResult ReverseJniEngine::onAudioReady(oboe::AudioStream *oboeStream,
+                                                        void *audioData,
+                                                        int32_t numFrames) {
+    JNIEnv *env;
+    // Attach the current thread to the JVM to make JNI calls.
+    int getEnvStat = mJavaVM->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    if (getEnvStat == JNI_EDETACHED) {
+        if (mJavaVM->AttachCurrentThread(&env, nullptr) != 0) {
+            __android_log_print(ANDROID_LOG_ERROR, TAG, "Failed to attach current thread to JVM");
+            return oboe::DataCallbackResult::Stop;
+        }
+    }
+
+    int xRunCount = mAudioStream->getXRunCount().value();
+
+    // Create a new Java float array to pass the audio data.
+    jfloatArray audioDataArray = env->NewFloatArray(numFrames * mAudioStream->getChannelCount());
+
+    // Call the Java method via reverse JNI.
+    env->CallVoidMethod(mJavaObject, mOnAudioReadyId, audioDataArray, numFrames, xRunCount);
+
+    if (env->ExceptionCheck()) {
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        __android_log_print(ANDROID_LOG_ERROR, TAG, "Exception thrown in Java onAudioReady");
+        return oboe::DataCallbackResult::Stop;
```

Comment:
There can be memory leak given the float array is not released.

Reply:
Thanks! Done